dissect/cstruct/types/enum.py

```diff
@@ -1,124 +1,179 @@
+        return list(map(cls, cls.type._read_array(stream, count, context)))
 
-    def __contains__(self, attr: str) -> bool:
-        return attr in self.values
+    def _read_0(cls, stream: BinaryIO, context: dict[str, Any] = None) -> list[Enum]:
+        return list(map(cls, cls.type._read_0(stream, context)))
 
-    def _read(self, stream: BinaryIO, context: dict[str, Any] = None) -> EnumInstance:
-        v = self.type._read(stream, context)
-        return self(v)
+    def _write(cls, stream: BinaryIO, data: Enum) -> int:
+        return cls.type._write(stream, data.value)
 
-    def _read_array(self, stream: BinaryIO, count: int, context: dict[str, Any] = None) -> List[EnumInstance]:
-        return list(map(self, self.type._read_array(stream, count, context)))
+    def _write_array(cls, stream: BinaryIO, array: list[Enum]) -> int:
+        data = [entry.value if isinstance(entry, Enum) else entry for entry in array]
+        return cls.type._write_array(stream, data)
 
-    def _read_0(self, stream: BinaryIO, context: dict[str, Any] = None) -> List[EnumInstance]:
-        return list(map(self, self.type._read_0(stream, context)))
+    def _write_0(cls, stream: BinaryIO, array: list[BaseType]) -> int:
+        data = [entry.value if isinstance(entry, Enum) else entry for entry in array]
+        return cls._write_array(stream, data + [cls.type()])
 
-    def _write(self, stream: BinaryIO, data: Union[int, EnumInstance]) -> int:
-        data = data.value if isinstance(data, EnumInstance) else data
-        return self.type._write(stream, data)
 
-    def _write_array(self, stream: BinaryIO, data: List[Union[int, EnumInstance]]) -> int:
-        data = [d.value if isinstance(d, EnumInstance) else d for d in data]
-        return self.type._write_array(stream, data)
+def _fix_alias_members(cls: type[Enum]) -> None:
+    # Emulate aenum NoAlias behaviour
+    # https://github.com/ethanfurman/aenum/blob/master/aenum/doc/aenum.rst
+    if len(cls._member_names_) == len(cls._member_map_):
+        return
 
-    def _write_0(self, stream: BinaryIO, data: List[Union[int, EnumInstance]]) -> int:
-        data = [d.value if isinstance(d, EnumInstance) else d for d in data]
-        return self.type._write_0(stream, data)
+    for name, member in cls._member_map_.items():
+        if name != member.name:
+            new_member = int.__new__(cls, member.value)
+            new_member._name_ = name
+            new_member._value_ = member.value
 
-    def default(self) -> EnumInstance:
-        return self(0)
+            type.__setattr__(cls, name, new_member)
+            cls._member_names_.append(name)
+            cls._member_map_[name] = new_member
+            cls._value2member_map_[member.value] = new_member
 
-    def default_array(self, count: int) -> List[EnumInstance]:
-        return [self.default() for _ in range(count)]
 
+class Enum(BaseType, IntEnum, metaclass=EnumMetaType):
```

Comment:
```suggestion
class Enum(IntEnum, BaseType, metaclass=EnumMetaType):
```

Is there a reason the first base class is not the Python built-in class here? All other cstruct types have the Python built-in as the first base class.

Reply:
This is illegal with Python's `Enum` class since it will think it's being used as a subclass:

```
dissect/cstruct/types/enum.py:98: in <module>
    class Enum(IntEnum, BaseType, metaclass=EnumMetaType):
/usr/lib64/python3.12/enum.py:508: in __prepare__
    member_type, first_enum = metacls._get_mixins_(cls, bases)
/usr/lib64/python3.12/enum.py:968: in _get_mixins_
    raise TypeError("new enumerations should be created as "
E   TypeError: new enumerations should be created as `EnumName([mixin_type, ...] [data_type,] enum_type)`
```